File: impl-large/zig/domain.zig
// Task Service, large tier — types, constants and pure rules.
const std = @import("std");
const http = std.http;
const Allocator = std.mem.Allocator;
const Value = std.json.Value;

pub const max_title_length = 80;
pub const max_name_length = 60;
pub const max_comment_length = 200;
pub const max_bulk_items = 20;
pub const min_priority = 1;
pub const max_priority = 5;
pub const default_limit = 20;
pub const max_limit = 100;
pub const default_quota = 10000;
pub const probe_quota = 5;
pub const port = 8080;

pub const roles = [_][]const u8{ "admin", "user" };
pub const statuses = [_][]const u8{ "todo", "in_progress", "done", "archived" };
pub const status_bonus = [_]i64{ 0, 3, 5, 0 };
pub const transitions = [_][2][]const u8{
    .{ "todo", "in_progress" }, .{ "todo", "archived" },  .{ "in_progress", "todo" },
    .{ "in_progress", "done" }, .{ "done", "archived" },
};
pub const project_sorts = [_][]const u8{ "id", "name", "taskCount" };
pub const task_sorts = [_][]const u8{ "id", "title", "priority", "score", "status" };
pub const user_sorts = [_][]const u8{ "id", "username", "role" };
pub const comment_sorts = [_][]const u8{ "id", "authorId" };
pub const seq_sorts = [_][]const u8{"seq"};
pub const group_bys = [_][]const u8{ "assignee", "status", "project" };

pub const User = struct {
    id: i64,
    username: []const u8,
    password: []const u8,
    role: []const u8,
    quota: i64,
    version: i64 = 1,
    deleted: bool = false,
};
pub const Session = struct { token: []const u8, userId: i64, used: i64 = 0 };
pub const Project = struct {
    id: i64,
    name: []const u8,
    ownerId: i64,
    version: i64 = 1,
    deleted: bool = false,
};
pub const Task = struct {
    id: i64,
    projectId: i64,
    title: []const u8,
    priority: i64,
    status: []const u8,
    assigneeId: ?i64,
    internalNote: []const u8 = "",
    version: i64 = 1,
    deleted: bool = false,
};
pub const Comment = struct { id: i64, taskId: i64, authorId: i64, body: []const u8 };
pub const AuditEntry = struct {
    seq: i64,
    actorId: i64,
    action: []const u8,
    resource: []const u8,
    resourceId: i64,
};
pub const OutboxEvent = struct {
    seq: i64,
    name: []const u8,
    resourceId: i64,
    delivered: bool = false,
};

pub const Detail = struct { field: []const u8, message: []const u8 };
pub const AppError = struct {
    status: http.Status,
    code: []const u8,
    message: []const u8,
    details: []const Detail,
};
pub const Error = error{App};

/// An error carries no payload in Zig, so the envelope waits here.
pub var app_error: AppError = undefined;

fn appError(
    status: http.Status,
    code: []const u8,
    message: []const u8,
    details: []const Detail,
) Error {
    app_error = .{ .status = status, .code = code, .message = message, .details = details };
    return error.App;
}

pub fn badRequest() Error {
    return appError(.bad_request, "bad_request", "the request is malformed", &.{});
}

pub fn unauthorized() Error {
    return appError(.unauthorized, "unauthorized", "authentication is required", &.{});
}

pub fn invalidCredentials() Error {
    return appError(.unauthorized, "invalid_credentials", "the username or password is wrong", &.{});
}

pub fn forbidden() Error {
    return appError(.forbidden, "forbidden", "you may not access this resource", &.{});
}

pub fn notFound() Error {
    return appError(.not_found, "not_found", "the resource does not exist", &.{});
}

pub fn conflict() Error {
    return appError(.conflict, "conflict", "the resource already exists", &.{});
}

pub fn invalidTransition() Error {
    return appError(.conflict, "invalid_transition", "the status change is not allowed", &.{});
}

pub fn preconditionFailed() Error {
    return appError(.precondition_failed, "precondition_failed", "the resource has changed", &.{});
}

pub fn preconditionRequired() Error {
    return appError(.precondition_required, "precondition_required", "the If-Match header is required", &.{});
}

pub fn quotaExceeded() Error {
    return appError(.too_many_requests, "quota_exceeded", "the request quota is exhausted", &.{});
}

pub fn invalid(details: []Detail) Error {
    std.mem.sort(Detail, details, {}, struct {
        fn lessThan(_: void, a: Detail, b: Detail) bool {
            const by_field = std.mem.order(u8, a.field, b.field);
            if (by_field != .eq) return by_field == .lt;
            return std.mem.order(u8, a.message, b.message) == .lt;
        }
    }.lessThan);
    return appError(.unprocessable_entity, "validation_failed", "the request body is not valid", details);
}

pub fn fail(field: []const u8, message: []const u8) Detail {
    return .{ .field = field, .message = message };
}

pub fn computeScore(priority: i64, status: []const u8) i64 {
    const base_score = priority * 10;
    for (statuses, status_bonus) |name, bonus| {
        if (std.mem.eql(u8, name, status)) return base_score + bonus;
    }
    return base_score;
}

pub fn isStatus(value: []const u8) bool {
    for (statuses) |name| {
        if (std.mem.eql(u8, name, value)) return true;
    }
    return false;
}

pub fn isRole(value: []const u8) bool {
    for (roles) |name| {
        if (std.mem.eql(u8, name, value)) return true;
    }
    return false;
}

pub fn isGroupBy(value: []const u8) bool {
    for (group_bys) |name| {
        if (std.mem.eql(u8, name, value)) return true;
    }
    return false;
}

pub fn allowedMove(from: []const u8, to: []const u8) bool {
    for (transitions) |move| {
        if (std.mem.eql(u8, move[0], from) and std.mem.eql(u8, move[1], to)) return true;
    }
    return false;
}

pub fn textLength(value: []const u8) usize {
    return std.unicode.utf8CountCodepoints(value) catch value.len;
}

pub fn containsIgnoreCase(haystack: []const u8, needle: []const u8) bool {
    if (needle.len > haystack.len) return false;
    var start: usize = 0;
    while (start + needle.len <= haystack.len) : (start += 1) {
        if (std.ascii.eqlIgnoreCase(haystack[start .. start + needle.len], needle)) return true;
    }
    return false;
}

pub fn checkString(
    value: []const u8,
    field: []const u8,
    max_length: usize,
    errors: *std.ArrayList(Detail),
    arena: Allocator,
) !void {
    if (value.len == 0) {
        const message = try std.fmt.allocPrint(arena, "{s} is required", .{field});
        try errors.append(arena, fail(field, message));
    } else if (textLength(value) > max_length) {
        const message = try std.fmt.allocPrint(arena, "{s} is too long", .{field});
        try errors.append(arena, fail(field, message));
    }
}

pub fn checkPriority(value: ?i64, errors: *std.ArrayList(Detail), arena: Allocator) !void {
    if (value == null or value.? < min_priority or value.? > max_priority) {
        try errors.append(arena, fail("priority", "priority is out of range"));
    }
}

pub fn checkStatus(value: Value, errors: *std.ArrayList(Detail), arena: Allocator) !void {
    const ok = switch (value) {
        .string => |text| isStatus(text),
        else => false,
    };
    if (!ok) try errors.append(arena, fail("status", "status is not valid"));
}

pub fn checkRole(value: Value, errors: *std.ArrayList(Detail), arena: Allocator) !void {
    const ok = switch (value) {
        .string => |text| isRole(text),
        else => false,
    };
    if (!ok) try errors.append(arena, fail("role", "role is not valid"));
}

pub fn checkQuota(value: Value, errors: *std.ArrayList(Detail), arena: Allocator) !void {
    const ok = switch (value) {
        .integer => |number| number >= 0,
        else => false,
    };
    if (!ok) try errors.append(arena, fail("quota", "quota is out of range"));
}

File: impl-large/rust/src/domain.rs
// Task Service, large tier — domain types, constants and pure rules.

use axum::http::StatusCode;
use axum::response::{IntoResponse, Response};
use axum::Json;
use serde_json::{json, Value};
use std::sync::{Arc, Mutex};

pub const MAX_TITLE_LENGTH: usize = 80;
pub const MAX_NAME_LENGTH: usize = 60;
pub const MAX_COMMENT_LENGTH: usize = 200;
pub const MAX_BULK_ITEMS: usize = 20;
pub const MIN_PRIORITY: i64 = 1;
pub const MAX_PRIORITY: i64 = 5;
pub const DEFAULT_LIMIT: i64 = 20;
pub const MAX_LIMIT: i64 = 100;
pub const DEFAULT_QUOTA: i64 = 10000;
pub const PROBE_QUOTA: i64 = 5;
pub const PORT: u16 = 8080;

pub const ROLES: [&str; 2] = ["admin", "user"];
pub const STATUSES: [&str; 4] = ["todo", "in_progress", "done", "archived"];
pub const STATUS_BONUS: [(&str, i64); 4] =
    [("todo", 0), ("in_progress", 3), ("done", 5), ("archived", 0)];
pub const TRANSITIONS: [(&str, &str); 5] = [
    ("todo", "in_progress"),
    ("todo", "archived"),
    ("in_progress", "todo"),
    ("in_progress", "done"),
    ("done", "archived"),
];
pub const PROJECT_SORTS: [&str; 3] = ["id", "name", "taskCount"];
pub const TASK_SORTS: [&str; 5] = ["id", "title", "priority", "score", "status"];
pub const USER_SORTS: [&str; 3] = ["id", "username", "role"];
pub const COMMENT_SORTS: [&str; 2] = ["id", "authorId"];
pub const SEQ_SORTS: [&str; 1] = ["seq"];
pub const GROUP_BYS: [&str; 3] = ["assignee", "status", "project"];

#[derive(Clone)]
pub struct User {
    pub id: i64,
    pub username: String,
    pub password: String,
    pub role: String,
    pub quota: i64,
    pub version: i64,
    pub deleted: bool,
}

#[derive(Clone)]
pub struct Session {
    pub user_id: i64,
    pub used: i64,
}

#[derive(Clone)]
pub struct Project {
    pub id: i64,
    pub name: String,
    pub owner_id: i64,
    pub version: i64,
    pub deleted: bool,
}

#[derive(Clone)]
pub struct Task {
    pub id: i64,
    pub project_id: i64,
    pub title: String,
    pub priority: i64,
    pub status: String,
    pub assignee_id: Option<i64>,
    pub internal_note: String,
    pub version: i64,
    pub deleted: bool,
}

#[derive(Clone)]
pub struct Comment {
    pub id: i64,
    pub task_id: i64,
    pub author_id: i64,
    pub body: String,
}

#[derive(Clone)]
pub struct AuditEntry {
    pub seq: i64,
    pub actor_id: i64,
    pub action: String,
    pub resource: String,
    pub resource_id: i64,
}

#[derive(Clone)]
pub struct OutboxEvent {
    pub seq: i64,
    pub name: String,
    pub resource_id: i64,
    pub delivered: bool,
}

/// What the handler tells the middleware to log and to place in the headers.
#[derive(Default)]
pub struct Out {
    pub user_id: Option<i64>,
    pub quota_remaining: Option<i64>,
    pub replayed: bool,
}

/// The per-request context that every layer reads.
#[derive(Clone)]
pub struct Ctx {
    pub request_id: String,
    pub token: String,
    pub key: Option<String>,
    pub if_match: Option<String>,
    pub out: Arc<Mutex<Out>>,
}

/// Every failure path returns this. `IntoResponse` turns it into the envelope.
pub struct AppError {
    pub status: StatusCode,
    pub code: &'static str,
    pub message: &'static str,
    pub request_id: String,
    pub details: Vec<Value>,
}

impl AppError {
    fn new(
        ctx: &Ctx,
        status: StatusCode,
        code: &'static str,
        message: &'static str,
        details: Vec<Value>,
    ) -> AppError {
        AppError {
            status,
            code,
            message,
            request_id: ctx.request_id.clone(),
            details,
        }
    }

    pub fn envelope(&self) -> Value {
        json!({ "error": {
            "code": self.code,
            "message": self.message,
            "requestId": self.request_id,
            "details": self.details,
        }})
    }
}

impl IntoResponse for AppError {
    fn into_response(self) -> Response {
        (self.status, Json(self.envelope())).into_response()
    }
}

macro_rules! failures {
    ($($name:ident, $status:ident, $code:literal, $message:literal;)*) => { $(
        pub fn $name(ctx: &Ctx) -> AppError {
            AppError::new(ctx, StatusCode::$status, $code, $message, vec![])
        }
    )* };
}

failures! {
    bad_request, BAD_REQUEST, "bad_request", "the request is malformed";
    unauthorized, UNAUTHORIZED, "unauthorized", "authentication is required";
    invalid_credentials, UNAUTHORIZED, "invalid_credentials", "the username or password is wrong";
    forbidden, FORBIDDEN, "forbidden", "you may not access this resource";
    not_found, NOT_FOUND, "not_found", "the resource does not exist";
    conflict, CONFLICT, "conflict", "the resource already exists";
    invalid_transition, CONFLICT, "invalid_transition", "the status change is not allowed";
    precondition_failed, PRECONDITION_FAILED, "precondition_failed", "the resource has changed";
    precondition_required, PRECONDITION_REQUIRED, "precondition_required",
        "the If-Match header is required";
    quota_exceeded, TOO_MANY_REQUESTS, "quota_exceeded", "the request quota is exhausted";
}

pub fn invalid(ctx: &Ctx, mut details: Vec<Value>) -> AppError {
    details.sort_by(|left, right| {
        (left["field"].as_str(), left["message"].as_str())
            .cmp(&(right["field"].as_str(), right["message"].as_str()))
    });
    AppError::new(
        ctx,
        StatusCode::UNPROCESSABLE_ENTITY,
        "validation_failed",
        "the request body is not valid",
        details,
    )
}

pub fn fail(field: &str, message: &str) -> Value {
    json!({ "field": field, "message": message })
}

pub fn compute_score(priority: i64, status: &str) -> i64 {
    let base_score = priority * 10;
    let bonus = STATUS_BONUS
        .iter()
        .find(|(name, _)| *name == status)
        .map_or(0, |(_, bonus)| *bonus);
    base_score + bonus
}

pub fn check_string(value: &str, field: &str, max_length: usize, errors: &mut Vec<Value>) {
    if value.is_empty() {
        errors.push(fail(field, &format!("{field} is required")));
    } else if value.chars().count() > max_length {
        errors.push(fail(field, &format!("{field} is too long")));
    }
}

pub fn check_priority(value: Option<i64>, errors: &mut Vec<Value>) {
    if !matches!(value, Some(number) if (MIN_PRIORITY..=MAX_PRIORITY).contains(&number)) {
        errors.push(fail("priority", "priority is out of range"));
    }
}

pub fn check_status(value: Option<&Value>, errors: &mut Vec<Value>) {
    if !value
        .and_then(Value::as_str)
        .is_some_and(|status| STATUSES.contains(&status))
    {
        errors.push(fail("status", "status is not valid"));
    }
}

pub fn check_role(value: &Value, errors: &mut Vec<Value>) {
    if !value.as_str().is_some_and(|role| ROLES.contains(&role)) {
        errors.push(fail("role", "role is not valid"));
    }
}

pub fn check_quota(value: &Value, errors: &mut Vec<Value>) {
    if !value.as_i64().is_some_and(|quota| quota >= 0) {
        errors.push(fail("quota", "quota is out of range"));
    }
}
